homeassistant/components/sensor/__init__.py

```diff
@@ -715,6 +739,37 @@ def state(self) -> Any:
                     f"but provides state {value}:{type(value)} resulting in '{err}'"
                 ) from err
 
+        # Enum checks
+        if value is not None and (
+            device_class == SensorDeviceClass.ENUM or self.options is not None
+        ):
+            if device_class != SensorDeviceClass.ENUM:
+                reason = "is missing the enum device class"
+                if device_class is not None:
+                    reason = f"has device class '{device_class}' instead of 'enum'"
+                raise ValueError(
+                    f"Sensor {self.entity_id} is providing enum options, but {reason}"
+                )
+
+            if state_class:
+                raise ValueError(
+                    f"Sensor {self.entity_id} has an state_class and thus indicating "
+                    "it has a numeric value; however, it has the enum device class"
+                )
+
+            if unit_of_measurement:
+                raise ValueError(
+                    f"Sensor {self.entity_id} has an unit of measurement and thus "
+                    "indicating it has a numeric value; "
+                    "however, it has the enum device class"
+                )
```

Comment:
These two I will extract and make them more generic in a separate PR, as these checks can be applied to all device classes that are marked as "not numeric".